File: cpp/gpusim-fe/gpusim-fe.Core/GridSimConfig/GridSimConfig.cpp
#include "GridSimConfig.h"
#include "Serialization/JavaXMLHelper.h"
#include "Serialization/JavaXMLListSerialize.hpp"

using namespace Core::GridSimConfig;
using namespace Core::GridSimConfig::Serialization;

//////////////////////////////////////////////////////////////////////////
// Constants
////////////////////////////////////////////////////////////////////////// 
#pragma region Public constants
const quint32 CGridSimConfig::c_minVersion           = 1;
const double  CGridSimConfig::c_minLinkBaudRate      = 1.0f;
const quint32 CGridSimConfig::c_currentConfigVersion = 1;
const double  CGridSimConfig::c_defaultLinkBaudRate  = 1.0f;
#pragma endregion

//////////////////////////////////////////////////////////////////////////

#pragma region Private constants
const QString CGridSimConfig::c_className            = QString("gpusim.config.GridSimConfig");
const QString CGridSimConfig::c_versionPropName      = QString("version");
const QString CGridSimConfig::c_linkBaudRatePropName = QString("linkBaudRate");
const QString CGridSimConfig::c_resourcesPropName    = QString("resources");
const QString CGridSimConfig::c_gridletsPropName     = QString("gridlets");
#pragma endregion

//////////////////////////////////////////////////////////////////////////
// Code
////////////////////////////////////////////////////////////////////////// 

#pragma region Constructor and equality comparision support
CGridSimConfig::CGridSimConfig(double linkBaudRate /*= c_defaultLinkBaudRate*/,
    const CGridSimResourcesConfig &resources /*= CGridSimResourcesConfig()*/,
    const CGridSimGridletsConfig &gridlets /*= CGridSimGridletsConfig()*/)
    : IJavaXMLSerialize(c_className), m_version(c_currentConfigVersion), m_linkBaudRate(c_defaultLinkBaudRate)
{
    setLinkBaudRate(linkBaudRate);
    setResources(resources);
    setGridlets(gridlets);
}

bool CGridSimConfig::operator==(const CGridSimConfig& other) const
{
    return
        (m_version   == other.m_version  ) &&
        (m_resources == other.m_resources) &&
        (m_gridlets  == other.m_gridlets );
}

bool CGridSimConfig::operator!=(const CGridSimConfig& other) const
{
    return !(*this == other);
}
#pragma endregion

//////////////////////////////////////////////////////////////////////////

#pragma region Properties
#pragma region Version property
quint32 CGridSimConfig::getVersion() const
{
    return m_version;
}

void CGridSimConfig::setVersion(quint32 version)
{
    if (m_version == version)
        return;

    Q_ASSERT(version >= c_minVersion);
    m_version = version;
}
#pragma endregion

#pragma region LinkBaudRate property
double CGridSimConfig::getLinkBaudRate() const
{
    return m_linkBaudRate;
}

void CGridSimConfig::setLinkBaudRate(double linkBaudRate)
{
    if (m_linkBaudRate == linkBaudRate)
        return;

    Q_ASSERT(linkBaudRate >= c_minLinkBaudRate);
    m_linkBaudRate = linkBaudRate;
}
#pragma endregion

#pragma region Resources property
CGridSimResourcesConfig CGridSimConfig::getResources() const
{
    return getResourcesRef();
}

const CGridSimResourcesConfig &CGridSimConfig::getResourcesRef() const
{
    return m_resources;
}

void CGridSimConfig::setResources(const CGridSimResourcesConfig &resources)
{
    if (m_resources == resources)
        return;

    Q_ASSERT(!resources.isEmpty());
    m_resources = resources;
}
#pragma endregion

#pragma region Gridlets property
CGridSimGridletsConfig CGridSimConfig::getGridlets() const
{
    return getGridletsRef();
}

const CGridSimGridletsConfig &CGridSimConfig::getGridletsRef() const
{
    return m_gridlets;
}

void CGridSimConfig::setGridlets(const CGridSimGridletsConfig &gridlets)
{
    if (m_gridlets == gridlets)
        return;

    Q_ASSERT(!gridlets.isEmpty());
    m_gridlets = gridlets;
}
#pragma endregion
#pragma endregion

//////////////////////////////////////////////////////////////////////////

#pragma region Serialization support
bool CGridSimConfig::isValid() const
{
    return (m_version >= c_minVersion) && (m_linkBaudRate >= c_minLinkBaudRate) &&
        (!(m_resources.isEmpty() || m_gridlets.isEmpty()));
}

void CGridSimConfig::saveDataToXMLNode(QDomDocument &doc, QDomElement &elem) const
{
    elem.appendChild(CJavaXMLHelper::createPropertyElement(doc, c_versionPropName,      getVersion()));
    elem.appendChild(CJavaXMLHelper::createPropertyElement(doc, c_linkBaudRatePropName, getLinkBaudRate()));

    elem.appendChild(createListPropertyElement(doc, c_resourcesPropName, getResourcesRef().constBegin(),
        getResourcesRef().constEnd()));
    elem.appendChild(createListPropertyElement(doc, c_gridletsPropName,  getGridletsRef().constBegin(),
        getGridletsRef().constEnd()));
}

void CGridSimConfig::loadDataFromXMLNode(const QDomElement &elem)
{
    quint32 version;
    double linkBaudRate;
    CGridSimResourcesConfig resources;
    CGridSimGridletsConfig gridlets;

    CJavaXMLHelper::getPropertyValue(elem, c_versionPropName,      version,      c_currentConfigVersion);
    CJavaXMLHelper::getPropertyValue(elem, c_linkBaudRatePropName, linkBaudRate, c_defaultLinkBaudRate);

    Serialization::getListPropertyValue(elem, c_resourcesPropName, std::back_inserter(resources));
    Serialization::getListPropertyValue(elem, c_gridletsPropName,  std::back_inserter(gridlets));

    setVersion(version);
    setLinkBaudRate(linkBaudRate);
    setResources(resources);
    setGridlets(gridlets);
}

IJavaXMLSerializePtr CGridSimConfig::create() const
{
    return IJavaXMLSerializePtr(new CGridSimConfig());
}

void CGridSimConfig::copy(const IJavaXMLSerializePtr &from)
{
    auto ptr = from.staticCast<CGridSimConfig>();
    *this = *ptr;
}
#pragma endregion

//////////////////////////////////////////////////////////////////////////

#pragma region Tools
CGridSimConfig CGridSimConfig::createTestConfig()
{
    // Machines list
    //
    CGridSimMachinesConfig machines;
    machines.append(CGridSimMachineConfig(6, 3, 5));

    //
    // Resources list
    //
    CGridSimResourceConfig resource("Test Architecture", "Test OS");
    resource.setBaudRate(100.0f);
    resource.setCostPerSec(55.0f);
    resource.setMachines(machines);
    resource.setCount(2);

    CGridSimResourcesConfig resources;
    resources.append(resource);

    // Gridlets list
    //
    CGridSimGridletsConfig gridltes;
    gridltes.append(CGridSimGridletConfig(100.0f, 50, 30, 10));

    // And finally config
    //
    return CGridSimConfig(560.0f, resources, gridltes);
}
#pragma endregion

File: java/gpusim-runtime/src/gpusim/config/GridSimConfig.java
package gpusim.config;

import java.io.Serializable;
import java.util.LinkedList;

public class GridSimConfig implements Serializable {
    //<editor-fold defaultstate="collapsed" desc="Constants">
    /**
     * Current target version of the GridSimConfig entity.
     */
    public static final int CURRENT_CONFIG_VERSION = 1;
    //</editor-fold>

    //<editor-fold defaultstate="collapsed" desc="Internal fields">
    private int _version;
    private double _linkBaudRate;
    private LinkedList<GridSimResourceConfig> _resources;
    private LinkedList<GridSimGridletConfig> _gridlets;
    //</editor-fold>

    //<editor-fold defaultstate="collapsed" desc="Constructors">
    public GridSimConfig() {
        _version = CURRENT_CONFIG_VERSION;
        _linkBaudRate = 1.0f;
        _resources = new LinkedList<GridSimResourceConfig>();
        _gridlets = new LinkedList<GridSimGridletConfig>();
    }
    //</editor-fold>

    //<editor-fold defaultstate="collapsed" desc="Properties">
    //<editor-fold defaultstate="collapsed" desc="Version property">
    public int getVersion() {
        return _version;
    }

    public void setVersion(int version) {
        if (version <= 0) {
            throw new AssertionError(version > 0);
        }

        _version = version;
    }
    //</editor-fold>

    //<editor-fold defaultstate="collapsed" desc="LinkBaudRate property">
    public double getLinkBaudRate() {
        return _linkBaudRate;
    }

    public void setLinkBaudRate(double linkBaudRate) {
        if (linkBaudRate < 1.0f) {
            throw new AssertionError(linkBaudRate >= 1.0f);
        }

        _linkBaudRate = linkBaudRate;
    }
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="Resources property">
    public LinkedList<GridSimResourceConfig> getResources() {
        return _resources;
    }

    public void setResources(LinkedList<GridSimResourceConfig> resources) {
        if (resources == null) {
            throw new AssertionError(resources != null);
        }

        _resources = resources;
    }
    //</editor-fold>

    //<editor-fold defaultstate="collapsed" desc="Gridlets property">
    public LinkedList<GridSimGridletConfig> getGridlets() {
        return _gridlets;
    }

    public void setGridlets(LinkedList<GridSimGridletConfig> gridlets) {
        if (gridlets == null) {
            throw new AssertionError(gridlets != null);
        }

        _gridlets = gridlets;
    }
    //</editor-fold>
    //</editor-fold>
}